front/lib/swr.ts

```diff
@@ -1028,6 +1042,83 @@ export function useWorkspaceEnterpriseConnection({
   };
 }
 
+type UseDataSourceKey = {
+  workspaceSid: string;
+  dataSourceName: string;
+  internalIds: string[];
+};
+
+export function useDataSourceNodes(
+  { workspaceSid, dataSourceName, internalIds }: UseDataSourceKey,
+  options?: SWRConfiguration<{
+    contentNodes: ContentNode[];
+    parentsById: Record<string, Set<string>>;
+  }>
+) {
+  const contentNodesFetcher: Fetcher<
+    { contentNodes: ContentNode[]; parentsById: Record<string, Set<string>> },
+    [string, string, string[]]
+  > = async ([workspaceSid, dataSourceName, internalIds]: [
+    string,
+    string,
+    string[]
+  ]) => {
+    if (internalIds.length === 0 || !dataSourceName) {
+      return { contentNodes: [], parentsById: {} };
+    }
+
+    const nodesUrl = `/api/w/${workspaceSid}/data_sources/${encodeURIComponent(
+      dataSourceName
+    )}/managed/content-nodes`;
+
+    const parentsUrl = `/api/w/${workspaceSid}/data_sources/${encodeURIComponent(
+      dataSourceName
+    )}/managed/parents`;
+
+    const [nodesData, parentsData]: [
+      GetContentNodesResponseBody,
+      GetContentNodeParentsResponseBody
+    ] = await Promise.all([
+      postFetcher([nodesUrl, { internalIds }]),
+      postFetcher([parentsUrl, { internalIds }]),
+    ]);
+
+    const { contentNodes } = nodesData;
+    if (contentNodes.length !== internalIds.length) {
+      throw new Error(
+        `Failed to fetch content nodes for all tables. Expected ${internalIds.length}, got ${contentNodes.length}.`
+      );
+    }
+
+    const parentsById = parentsData.nodes.reduce((acc, r) => {
+      acc[r.internalId] = new Set(r.parents);
+      return acc;
+    }, {} as Record<string, Set<string>>);
+
+    return { contentNodes, parentsById };
+  };
+
+  const { mutate: globalMutate } = useSWRConfig();
+
+  const preload = (
+    internalIds: string[],
+    contentNodes: ContentNode[],
+    parentsById: Record<string, Set<string>>
+  ) => {
+    void globalMutate([workspaceSid, dataSourceName, internalIds], {
```

Comment:
Seen IRL, let's avoid a global mutate here and implement a proper optimistic update that can be rollbacked if the server fails.